### PR TITLE
[D-326] Store accessor support for assignable_values

### PR DESCRIPTION
**Pull request by begerdom**

### CODE REVIEW

File: spec/assignable_values/active_record_spec.rb [hunk 1]
```diff
@@ -370,6 +370,211 @@ def save_without_validation(record)
...
+          @klass.new(:format => 'disallowed value').should_not be_valid
+        end
+
+        it 'should use the same error message as validates_inclusion_of' do
+          record = @klass.new(:format => 'disallowed value')
+          record.valid?
+          errors = record.errors[:format]
+          error = errors.respond_to?(:first) ? errors.first : errors # the return value sometimes was a string, sometimes an Array in Rails
+          error.should == I18n.t('errors.messages.inclusion')
+          error.should == 'is not included in the list'
```
Comment (FLeinzi) [inline spec/assignable_values/active_record_spec.rb:398]: What's the difference?

File: spec/assignable_values/active_record_spec.rb [hunk 2]
```diff
@@ -370,6 +370,211 @@ def save_without_validation(record)
...
+        it 'should allow a previously saved, blank format value even if that value is no longer allowed' do
+          record = @klass.create!(:format => 'mp3')
+
+          record.update_column(:metadata, { 'format' => nil }) # update without validations for the sake of this test
+          record.reload.should be_valid
+        end
+
+        it 'should not allow nil (the "previous value") if the record was never saved' do
+          record = @klass.new(:format => nil)
+          # Show that nil is not assignable, even though `record.genre_was` is nil.
```
Comment (FLeinzi) [inline spec/assignable_values/active_record_spec.rb:421]: Comment does not match to the example.

File: spec/assignable_values/active_record_spec.rb [hunk 3]
```diff
@@ -370,6 +370,211 @@ def save_without_validation(record)
...
+          before :each do
+            @klass = Song.disposable_copy do
+              store :metadata, accessors: [:instruments], coder: JSON
+
+              assignable_values_for :instruments, multiple: true do
+                %w[piano drums guitar]
+              end
+            end
+          end
+
```
Comment (FLeinzi) [inline spec/assignable_values/active_record_spec.rb:457]: No test for multiple assignment?

**Review by FLeinzi — APPROVED** [review]
LGTM!

I don't have any opinion to that suggestion:

> As store attribute are its own kind of attribute (like belongs_to relations), they should probably be handled in an own restriction class, like AssignableValues::ActiveRecord::Restriction::StoreAttribute to explicitly handle all peculiarities of store attributes.

For me this change would be enough as store accessors are similar to scalar attributes.

**Comment by triskweline** [comment]
> As store attribute are its own kind of attribute (like belongs_to relations), they should probably be handled in an own restriction class, like AssignableValues::ActiveRecord::Restriction::StoreAttribute to explicitly handle all peculiarities of store attributes.

The original gem design used multiple restriction classes to prevent the `if`/`else` clauses seen in this diff. The reader already needs to understand a lot of cases for covering various Rails versions.

If you find the restriction to be similiar to `ScalarAttribute`, you can inherit or extract a common module.

File: spec/assignable_values/active_record_spec.rb (hunk 1 above)
Comment (begerdom) [reply spec/assignable_values/active_record_spec.rb]: C&P thing. Adapted the original line as well

**Pull request merged**